### ISSUE TITLE
Support roundtripping new features of MHLO to/from StableHLO

### ISSUE THREAD

**Issue by burmako** [openxla/stablehlo#593]
#447 provides a good illustration for the workflow of adding new features. First, there was [an RFC](https://github.com/openxla/xla/discussions/22). Then, there was [an implementation in MHLO](https://github.com/tensorflow/mlir-hlo/commit/f9d277f903d20ca4a5af20c9088d9ce413d33dbc). Finally, if/when the RFC is accepted, we'll add the proposed functionality to StableHLO.

Enabling this workflow was the whole point of splitting StableHLO and MHLO. This way, new features can be introduced in MHLO relatively quickly, becoming immediately available to producers/consumers without being blocked on being specced / evaluated for inclusion in StableHLO.

We are aiming for StableHLO to be adopted by a wide variety of ML frameworks including TensorFlow (#216), JAX (#217) and PyTorch (#218 and #219), and one prerequisite for that is figuring out how to allow these frameworks to access these MHLO features.

In the spirit of MLIR, a natural solution to this problem involves frameworks producing a mixture of StableHLO and MHLO dialects, but that doesn't solve the problem of serialization and compatibility, because this is not something that we're planning to prioritize in the MHLO dialect (the goal of the MHLO dialect is to provide maximum flexibility for compiler authors and compatibility guarantees get in the way).

I think there's a fairly straightforward way of resolving this. We should change MHLO to StableHLO roundtrip, i.e. [hlo_legalize_to_stablehlo](https://github.com/tensorflow/mlir-hlo/tree/master/mhlo/transforms/hlo_legalize_to_stablehlo) + [stablehlo_legalize_to_hlo](https://github.com/tensorflow/mlir-hlo/tree/master/mhlo/transforms/stablehlo_legalize_to_hlo), to roundtrip these MHLO features through `stablehlo.custom_call` instead of [rejecting them](https://github.com/tensorflow/mlir-hlo/blob/723f86de68599210fc3f85b231ee86783f9e9c1b/mhlo/transforms/hlo_legalize_to_stablehlo/hlo_legalize_to_stablehlo.cc#L163-L167). This way, frameworks can produce a mixture of StableHLO and MHLO, then run `hlo-legalize-to-stablehlo` and voila - they end up with an artifact that has compatibility guarantees.

**Comment by burmako** [comment]
Implemented initial support for roundtripping in https://github.com/openxla/xla/pull/159.

The PR linked above provides initial support for roundtripping, leaving a few follow-up features to follow-up PRs:
  1. The extensibility protocol that automatically packs MHLO ops into stablehlo.custom_call and unpacks them back into MHLO doesn't support regions yet. This functionality isn't needed at the moment, and its absence is not a blocker (manual extensibility protocol - i.e. modifying hlo_legalize_to_stablehlo + stablehlo_legalize_to_hlo by hand - can be employed instead if needed), so it is left for future work.
  2. Roundtripping MHLO-specific types is not supported yet. I think we should extend StableHLO with !stablehlo.custom_type that should solve this problem, but that will need to happen in an RFC. I've opened a ticket to track this: #743.
